magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

KrasnoshchokBohdan opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->

<details>
  <summary>createCompareList</summary>

```graphql
mutation {
  createCompareList(
    input: {
      products: ["1", "2"]
    }
  ) {
    uid
    item_count
    attributes {
      code
      label
    }
    items {
      uid
      product {
        sku
        name
        description {
          html
        }
      }
    }
  }
}
```
</details>

<details>
  <summary>addProductsToCompareList</summary>

```graphql
mutation {
  addProductsToCompareList(
    input: { uid: "-", products: ["3"] }
  ) {
    uid
    item_count
    attributes {
      code
      label
    }
    items {
      uid
      product {
        sku
        name
        description {
          html
        }
      }
    }
  }
}

```
</details>

<details>
  <summary>removeProductsFromCompareList</summary>

```graphql
mutation {
  removeProductsFromCompareList(
    input: { uid: "-", products: ["2"] }
  ) {
    uid
    item_count
    attributes {
      code
      label
    }
    items {
      uid
      product {
        sku
        name
        description {
          html
        }
      }
    }
  }
}

```
</details>
 
After these mutations, the counter in the header and in the customer account does not change. I found \Magento\Catalog\Helper\Product\Compare::calculate, which updates the quantity in _catalogSession, but the counters actually get their data from the customer data (compare-products section). To get an accurate quantity, the session needs to be updated, which is quite contradictory when considering the issue in the GraphQL scope.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#[39905](https://github.com/magento/magento2/issues/39905)

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Use any GraphQL mutation from the PR description.
2. Log in as a customer and go to /catalog/product_compare/index/
3. Go to /customer/account/index/ and check the "Compare Products" section.

<img width="259" alt="image" src="https://github.com/user-attachments/assets/d00bd4fc-aeae-4842-9c66-71706eb9648a" />


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)
 - [ ]

KrasnoshchokBohdan committed:
magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

Integrates the Product Compare helper across compare list resolvers to trigger updates after adding, removing, or creating compare lists. Also forces a reload of compare product data on the frontend to maintain accuracy.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39905

KrasnoshchokBohdan commented:
@magento run all tests

Den4ik reviewed (changes requested):
Hi @KrasnoshchokBohdan 
Thanks for your PR.
Could you please fix static tests and add new tests for fixed issue?
Thanks

KrasnoshchokBohdan commented on app/code/Magento/Catalog/view/frontend/web/js/view/compare-products.js:
@Den4ik Hi!
Sorry to bother you. I probably should have marked this as a draft immediately.
Currently, I'm struggling with a solution for reloading the compare-product customer section after changes made by a GraphQL mutation (

Den4ik replied:
Hi @KrasnoshchokBohdan 
Thanks for details

KrasnoshchokBohdan committed:
magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

Introduced CompareCookieManager to manage cookie invalidation for compare list products. Updated resolvers to use this service and ensure cookies are marked invalid after product compare list operations. Removed redundant client-side cookie reload logic in the frontend.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39905

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39905

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

Introduced `CompareCookieManager` enhancements with better cookie management through visibility modifiers and documentation changes. Suppressed warnings for PHPMD where relevant and added comprehensive unit tests for cookie handling, including edge cases like size limits and failures.

KrasnoshchokBohdan committed:
magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

- Static tests fix

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39905

KrasnoshchokBohdan commented:
@Den4ik, the only solution I found for now is to change compare-products cookie, so if you have any comments or suggestions, I will be happy to consider them

Den4ik commented:
Hi @KrasnoshchokBohdan 
I don't mind using cookies manager, but want to check it locally.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39905

engcom-Hotel commented on app/code/Magento/CompareListGraphQl/Model/Service/CompareCookieManager.php:
```diff
@@ -0,0 +1,106 @@
+     * @param CookieMetadataFactory $cookieMetadataFactory
+     * @param LoggerInterface $logger
+     */
+    public function __construct(
```
Make readonly constructor parameters.

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the contribution!

This is good to have PR. but please have a look into the below review comment.

Thanks

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-39905

KrasnoshchokBohdan committed:
magento/magento2#39905 Product Removed on Mobile Still Appears in Web's Mini Compare Section Until Re-login

Refactor CompareCookieManager to use readonly properties

Replaced manually assigned properties with readonly properties in the constructor to simplify initialization and improve immutability. Updated exception handling to use the imported Exception class instead of a fully qualified name.

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-39905

engcom-Hotel reviewed (approved):
Failed tests seems to flaky, hence approving the PR.

engcom-Bravo committed:
Merge branch '2.4-develop' into fix-for-issue-39905

engcom-Bravo commented:
@magento run all tests

engcom-Charlie commented:
@magento run Functional Tests B2B
